src/databricks/labs/ucx/source_code/dfsa.py

```diff
@@ -0,0 +1,116 @@
+import logging
+from collections.abc import Iterable
+from dataclasses import dataclass
+from pathlib import Path
+
+from sqlglot import Expression as SqlExpression, parse as parse_sql, ParseError as SqlParseError
+from sqlglot.expressions import Identifier, Literal
+
+from databricks.sdk.service.workspace import Language
+
+from databricks.labs.ucx.source_code.base import (
+    is_a_notebook,
+    CurrentSessionState,
+    file_language,
+    guess_encoding,
+    DIRECT_FS_REFS,
+)
+from databricks.labs.ucx.source_code.graph import DependencyGraph, Dependency
+from databricks.labs.ucx.source_code.notebooks.sources import Notebook
+from databricks.labs.ucx.source_code.path_lookup import PathLookup
+from databricks.labs.ucx.source_code.queries import FromTable
+
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class DFSA:
+    """A DFSA is a record describing a Direct File System Access"""
+
+    path: str
+
+
+class DfsaCollector:
```

Comment:
Things get a lot more complicated once we deal with not only sql code but also python code and workspace queries (which is anticipated here). See full-fledged collector here https://github.com/databrickslabs/ucx/blob/e62eaf1e4113179016183dc177b037b842f5cefe/src/databricks/labs/ucx/source_code/dfsa.py
That said, there's definitely redundant code for walking dependency graphs. Maybe we should introduce a DependencyGraphWalker to avoid that ?
Re caching, wouldn't it make sense to implement it as part of (or on top of) WorkspacePath rather than retrofitting existing application code?

Reply:
we are rate limited here and we cannot afford 2x10hrs to fetch notebook contents.